File: python/easy/500-keyboard-row.py
#!/usr/bin/env python
#coding: utf-8


class Solution(object):
    def findWords(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        row1 = 'qwertyuiopQWERTYUIOP'
        row2 = 'asdfghjklASDFGHJKL'
        row3 = 'zxcvbnmZXCVBNM'

        results = []
        for word in words:
            num = None
            for c in word:
                if c in row1:
                    n = 1
                elif c in row2:
                    n = 2
                elif c in row3:
                    n = 3

                if num is None:
                    num = n
                elif num != n:
                    num = None
                    break

            if num is not None:
                results.append(word)

        return results


def main():
    cases = [
        (["Hello", "Alaska", "Dad", "Peace"], ["Alaska", "Dad"])
    ]
    for inputs, outputs in cases:
        results = Solution().findWords(inputs)
        print 'Results', results
        assert results == outputs

if __name__ == '__main__':
    main()

File: rust/easy/500-keyboard-row.rs

const ROW1: &'static str = "qwertyuiopQWERTYUIOP";
const ROW2: &'static str = "asdfghjklASDFGHJKL";
const ROW3: &'static str = "zxcvbnmZXCVBNM";

fn find_words<'a>(words: &Vec<&'a str>) -> Vec<&'a str> {
    let mut results = vec![];
    for word in words {
        let mut num: Option<usize> = None;
        for c in word.chars() {
            let n = if ROW1.find(c).is_some() {
                Some(1)
            } else if ROW2.find(c).is_some() {
                Some(2)
            } else if ROW3.find(c).is_some() {
                Some(3)
            } else {
                None
            };
            if num.is_none() {
                num = n;
            } else if num != n {
                num = None;
                break;
            }
        }
        if num.is_some() {
            results.push(*word);
        }
    }
    results
}

fn main() {
    for (inputs, mut outputs) in vec![
        (vec!["Hello", "Alaska", "Dad", "Peace"], vec!["Alaska", "Dad"])
    ] {
        let results = find_words(&inputs);
        assert_eq!(results.len(), outputs.len());
        outputs.sort();
        for word in results {
            assert_eq!(outputs.binary_search(&word).is_ok(), true);
            println!("Row(ok): {}", word);
        }
        println!("All OK!!!")
    }
}
